validation: Give `m_block_index` ownership of `CBlockIndex`s

dongcarl opened the pull request:
Part of: #24303
Split off from: #22564

```
Instead of having CBlockIndex's live on the heap, which requires manual
memory management, have them be owned by m_block_index. This means that
they will live and die with BlockManager.
```

The second commit demonstrates how this makes calls to `Unload()` to satisfy the address sanitizer unnecessary.

dongcarl committed:
blockstorage: Make m_block_index own CBlockIndex's

Instead of having CBlockIndex's live on the heap, which requires manual
memory management, have them be owned by m_block_index. This means that
they will live and die with BlockManager.

A change to BlockManager::LookupBlockIndex:
- Previously, it was a const member function returning a non-const CBlockIndex*
- Now, there's are const and non-const versions of
  BlockManager::LookupBlockIndex returning a CBlockIndex with the same
  const-ness as the member function:
    (e.g. const CBlockIndex* LookupBlockIndex(...) const)

See next commit for some weirdness that this eliminates.

The range based for-loops are modernize (using auto + destructuring) in
a future commit.

dongcarl committed:
tests: Remove now-unnecessary manual Unload's

These manual calls to Unload() are no longer necessary because
CBlockIndex's no longer live in the heap as of the previous commit.

MarcoFalke commented on src/validation.cpp:
```diff
@@ -4181,8 +4181,8 @@ void CChainState::CheckBlockIndex()
     std::multimap<CBlockIndex*,CBlockIndex*> forward;
-    for (const std::pair<const uint256, CBlockIndex*>& entry : m_blockman.m_block_index) {
-        forward.insert(std::make_pair(entry.second->pprev, entry.second));
+    for (std::pair<const uint256, CBlockIndex>& entry : m_blockman.m_block_index) {
+        forward.insert(std::make_pair(entry.second.pprev, &entry.second));
```
nit: (assuming it compiles)

```suggestion
    for (const auto& [_, block] : m_blockman.m_block_index) {
        forward.emplace(block.pprev, &block));
```

MarcoFalke commented on src/validation.cpp:
```diff
@@ -2965,8 +2965,8 @@ bool CChainState::InvalidateBlock(BlockValidationState& state, CBlockIndex* pind
-        for (const auto& entry : m_blockman.m_block_index) {
-            CBlockIndex *candidate = entry.second;
+        for (auto& entry : m_blockman.m_block_index) {
+            CBlockIndex *candidate = &entry.second;
```
nit: clang-format touched code

MarcoFalke commented on src/node/blockstorage.h:
```diff
@@ -49,7 +50,11 @@ extern bool fPruneMode;
+// we ever switch to another associative container, we need to either use a
+// container that has stable addressing (true of all std associative
+// containers), or make the key a `std::unique_ptr<CBlockIndex>`
+typedef std::unordered_map<uint256, CBlockIndex, BlockHasher> BlockMap;
```
```suggestion
using BlockMap = std::unordered_map<uint256, CBlockIndex, BlockHasher>;
```

nit

maflcko reviewed (commented):
Left some nits. Feel free to ignore any or all of them.

ajtowns commented on src/node/blockstorage.cpp:
```diff
@@ -202,12 +204,13 @@ CBlockIndex* BlockManager::InsertBlockIndex(const uint256& hash)
     // Create new
-    CBlockIndex* pindexNew = new CBlockIndex();
-    mi = m_block_index.insert(std::make_pair(hash, pindexNew)).first;
+    CBlockIndex new_index{};
+    mi = m_block_index.insert(std::make_pair(hash, std::move(new_index))).first;
```
I think you could replace both that and the previous block with `try_emplace` ?

```c++
    // Return existing or create new
    auto [mi, inserted] = m_block_index.try_emplace(hash);
    CBlockIndex* pindex = &(*mi).second;
    if (inserted) {
        pindex->phashBlock = &((*mi).first);
    }
    return pindex;
```

dongcarl replied:
Much nicer! Committed!

dongcarl committed:
refactor: Rewrite InsertBlockIndex with try_emplace

Credit to ajtowns for this suggestion, thanks!

dongcarl committed:
style-only: Use using instead of typedef for BlockMap

dongcarl committed:
style: Modernize range-based loops over m_block_index

dongcarl commented:
Okay I'm still reserving my right to ignore future nits, but these one struck a soft spot in my heart so I incorporated them 😁

jamesob reviewed (approved):
ACK 8fb03ca47c01907b98461b8d538e2691d7ef65af ([`jamesob/ackr/24050.1.dongcarl.validation_move_cblockin`](https://github.com/jamesob/bitcoin/tree/ackr/24050.1.dongcarl.validation_move_cblockin))

Good change! For what it's worth, my previous benchmarking may be a relevant consideration for reviewers: https://github.com/bitcoin/bitcoin/pull/22564#issuecomment-930377301

Some of the intermediate commits could probably be squashed, but that doesn't bother me much either way.

I've compiled each commit and run unittests locally.

<details><summary>Show signature data</summary>
<p>

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

ACK 8fb03ca47c01907b98461b8d538e2691d7ef65af ([`jamesob/ackr/24050.1.dongcarl.validation_move_cblockin`](https://github.com/jamesob/bitcoin/tree/ackr/24050.1.dongcarl.validation_move_cblockin))

Good change! For what it's worth, my previous benchmarking may be a relevant consideration for reviewers: https://github.com/bitcoin/bitcoin/pull/22564#issuecomment-930377301

Some of the intermediate commits could probably be squashed, but that doesn't bother me much either way.

I've compiled each commit and run unittests locally.
-----BEGIN PGP SIGNATURE-----

iQIzBAEBCgAdFiEEGNRVI1NPYuZCSIrGepNdrbLETwUFAmHnJrIACgkQepNdrbLE
TwWLJhAAi3IfNFNZIaqTnTfMw8V9W8v2K2wNXT29tsj09JVTaJ5PGpQ9WoBbrkP+
cTtTvmlJ71X9VvRERPL0oz8A4YF1/QOPj9NJ+ZVE3NLg3z6rTbn5PFCqBNUYUvp/
a+qLshzpCsluQKs7B+70qpu0L589yYEcO6cKHVRVPLi8jrJxCc7/lEo2GdCjD2xJ
nxj0Pu+O+erN7zhVZ4evqOr6j+XSufHn6xiqm7DdGXzyafQzpPB3h5FJxVPZbh5p
PI5plMf5WcxPePd8ynE5v/wnKCZkaMhJm1cNC9D/MjmJUIPZyx8ngJoKPllgztHf
UzKZ0n7MBPHwFdsMGraKq0VCotD/ZxFGbeNtXaI0BLZfjA2Xyf1vaFFfNRg5uZHe
d45/Sa8uDQjVApxQtIbfRMGNFVms5pr8SrNuPvJ0EDIXtYXXuEvkVmX5lhkz39Sy
GO4/IbqBiT7md+dk7J7h5M6j3C5LnyiFWJzqS418//kk24/KSF6/kwGzXp1ASweY
H85Efa9xq8dr1KBiSw9wZ0U3RFpqOyz+/0ejzrBVxtMjd9x49RF4MRmakNPi/P4D
+SVBS0e0gr/FV0CW501tc5gZYMC3UK+whGcmaCT0LRjjQlpFv0M1sZZbA6CGdFYQ
zFR7x0aXcIZjByDECGMi9+T6c3Mi0yGa2ulc8uO7vo3DNLuTJOQ=
=f/Ec
-----END PGP SIGNATURE-----

```

</p></details>

<details><summary>Show platform data</summary>
<p>

```
Tested on Linux-5.10.0-10-amd64-x86_64-with-glibc2.31

Configured with ./configure LDFLAGS=-L/home/james/src/bitcoin/db4/lib/ CPPFLAGS=-I/home/james/src/bitcoin/db4/include/ CXXFLAGS=-fPIE -pipe -O2 -g -Wthread-safety-analysis -Wall -Werror=sign-compare -Wsign-compare -Werror=thread-safety-analysis --enable-wallet --enable-debug --with-daemon --enable-natpmp-default CXX=/usr/local/bin/clang++ CC=/usr/local/bin/clang --disable-shared --with-pic --enable-benchmark=no --enable-module-recovery --enable-module-schnorrsig --enable-experimental PYTHONPATH= --disable-shared --with-pic --enable-benchmark=no --enable-module-recovery --enable-module-schnorrsig --enable-experimental --no-create --no-recursion

Compiled with /usr/bin/ccache /usr/local/bin/clang++ -std=c++17 -mavx -mavx2 -mpclmul -fPIE -pipe -O2 -g -Wthread-safety-analysis -Wall -Werror=sign-compare -Wsign-compare -Werror=thread-safety-analysis -O0 -g3 -ftrapv -fdebug-prefix-map=$(abs_top_srcdir)=.  -Wstack-protector -fstack-protector-all -fcf-protection=full -fstack-clash-protection -msse4 -msha -msse4.1 -msse4.2  i

Compiler version: Debian clang version 13.0.1-++20220112082946+b9a243d1cac2-1~exp1~20220112203027.57
```

</p></details>

ajtowns commented on src/validation.cpp:
```diff
@@ -3089,17 +3089,17 @@ void CChainState::ResetBlockFailureFlags(CBlockIndex *pindex) {
     // Remove the invalidity flag from this block and all its descendants.
     BlockMap::iterator it = m_blockman.m_block_index.begin();
     while (it != m_blockman.m_block_index.end()) {
```
Could be a `for` loop if you're modernising things. Ditto above.

ajtowns commented on src/validation.cpp:
```diff
@@ -3428,7 +3428,7 @@ bool ChainstateManager::AcceptBlockHeader(const CBlockHeader& block, BlockValida
         if (miSelf != m_blockman.m_block_index.end()) {
             // Block header is already known.
-            CBlockIndex* pindex = miSelf->second;
+            CBlockIndex* pindex = &(miSelf->second);
```
Adding brackets to distinguish from `(&miSelf)->second` doesn't seem that useful to me...

ajtowns reviewed (approved):
ACK 8fb03ca47c01907b98461b8d538e2691d7ef65af ; code review only. some more nits you may want to consider

MarcoFalke commented:
For reference, the stuff is still on the heap. This only changes the layout a bit.

Master:

![Screenshot from 2022-01-19 18-25-45](https://user-images.githubusercontent.com/6399679/150183454-5ea149c8-34fc-4d34-a0ef-41fa8045d5b3.png)


This:

![Screenshot from 2022-01-19 18-27-02](https://user-images.githubusercontent.com/6399679/150183490-dafe809f-aad4-45bc-91c5-d68c72d55829.png)

dongcarl committed:
refactor: Rewrite AddToBlockIndex with try_emplace

dongcarl commented:
Pushed 8fb03ca47c01907b98461b8d538e2691d7ef65af -> c68516d662687699b399f8d12e22f2e985edae1a
- Some more modernizations

-----

Anything else this needs before it's ready for merge?

MarcoFalke commented on src/wallet/test/wallet_tests.cpp:
```diff
@@ -362,10 +362,10 @@ static int64_t AddTx(ChainstateManager& chainman, CWallet& wallet, uint32_t lock
     if (blockTime > 0) {
         LOCK(cs_main);
-        auto inserted = chainman.BlockIndex().emplace(GetRandHash(), new CBlockIndex);
+        auto inserted = chainman.BlockIndex().emplace(std::piecewise_construct, std::make_tuple(GetRandHash()), std::make_tuple());
```
first commit: Is there any reason/advantage to change this to std::piecewise_construct and not try_emplace, which automatically enables that. Or would that not work because the `CBlockIndex{}` rvalue can't be elided?

```cpp
auto inserted{chainman.BlockIndex().try_emplace(GetRandHash(), CBlockIndex{})};

MarcoFalke commented on src/node/blockstorage.cpp:
```diff
@@ -199,18 +196,13 @@ CBlockIndex* BlockManager::InsertBlockIndex(const uint256& hash)
-    BlockMap::iterator mi = m_block_index.find(hash);
-    if (mi != m_block_index.end()) {
-        return (*mi).second;
+    // Return existing or create new
+    auto [mi, inserted] = m_block_index.try_emplace(hash);
```
```suggestion
    const auto [mi, inserted]{m_block_index.try_emplace(hash)};
```

In the third commit: I think try_emplace is self-explanatory and the comment can be removed

MarcoFalke commented on src/node/blockstorage.cpp:
```diff
@@ -199,18 +196,13 @@ CBlockIndex* BlockManager::InsertBlockIndex(const uint256& hash)
-        return (*mi).second;
+    // Return existing or create new
+    auto [mi, inserted] = m_block_index.try_emplace(hash);
+    CBlockIndex* pindex = &(*mi).second;
```
in the third commit: Any reason to use a pointer here when a reference is sufficient?

```diff
diff --git a/src/node/blockstorage.cpp b/src/node/blockstorage.cpp
index 41623186ff..cab8c20e53 100644
--- a/src/node/blockstorage.cpp
+++ b/src/node/blockstorage.cpp
@@ -203,11 +203,11 @@ CBlockIndex* BlockManager::InsertBlockIndex(const uint256& hash)
 
     // Return existing or create new
     auto [mi, inserted] = m_block_index.try_emplace(hash);
-    CBlockIndex* pindex = &(*mi).second;
+    CBlockIndex& block{(*mi).second};
     if (inserted) {
-        pindex->phashBlock = &((*mi).first);
+        block.phashBlock = &((*mi).first);
     }
-    return pindex;
+    return &block;
 }
 
 bool BlockManager::LoadBlockIndex(

MarcoFalke commented on src/node/blockstorage.cpp:
```diff
@@ -224,8 +216,8 @@ bool BlockManager::LoadBlockIndex(
-    for (const std::pair<const uint256, CBlockIndex*>& item : m_block_index) {
-        CBlockIndex* pindex = item.second;
+    for (auto& [_, block_index] : m_block_index) {
+        CBlockIndex* pindex = &block_index;
         vSortedByHeight.push_back(std::make_pair(pindex->nHeight, pindex));
```
nit in the for-loop commit: Any reason to keep this alias?

```suggestion
        vSortedByHeight.emplace_back(block_index.nHeight, &block_index));
```

MarcoFalke commented on src/node/blockstorage.cpp:
```diff
@@ -386,8 +374,8 @@ bool BlockManager::LoadBlockIndexDB(ChainstateManager& chainman)
-    for (const std::pair<const uint256, CBlockIndex*>& item : m_block_index) {
-        CBlockIndex* pindex = item.second;
+    for (const auto& [_, block_index] : m_block_index) {
+        const CBlockIndex* pindex = &block_index;
```
same

maflcko reviewed (approved):
left some style nits. Feel fee to ignore any or all. I am happy to re-ACK if you don't.

review ACK c68516d662687699b399f8d12e22f2e985edae1a 🍏

<details><summary>Show signature</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

review ACK c68516d662687699b399f8d12e22f2e985edae1a 🍏
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUiTPgv8CbaBOK51qXPuhdeM7/DZJAMYFCAPOYAsAHSD7mrO1PG/7wA5vUW2ffGY
U2yM/rLCnBBCITrcrOygU61k72HPwgGTQ7jG0zV5+A339cAbnpJBm5+O8ky5/+NJ
/88BlcGNOmmZwUnuRLOwWwILDvNKPb4SwwBYSEDCWVS3hWt8e26Kw08xwiemHKsG
jyAIUZy8o789VNw5Wt12UQLbTZ/6p0Ox77le7egFttnC/Bpea0IbOdCTZzWOCiZQ
PUGcl+YcXMZxEnV1n+3GWWWMQgOYRYLa4BGAPVeauQd3AWEIU1jwfGL3Sy0Mik24
3Ej/Qx9UCMMhv7DB96nBeaYWoY7txJhT6c2ZnmseggFuKdtxcvIpd/nWEsbdEH4S
3Tn0ZZHSFXhPA76ChmS8eHYid7NIxkXPff43bCAz8cERJ68ISm061gqi6RaD0Rxl
sWOrxIzJpDg0caTkaHo8p2ZPVqm+vaPsoxlBJiYoAGKvQzXz/EjzOazgPbEqk3ZK
5MZUwBRK
=Ofkd
-----END PGP SIGNATURE-----
```


</details>

dongcarl commented:
Sorry, not going to make any more style/modernization-related changes, they can be followups.

jamesob commented on src/node/blockstorage.h:
```diff
@@ -49,7 +50,11 @@ extern bool fPruneMode;
+// Because validation code takes pointers to the map's CBlockIndex objects, if
+// we ever switch to another associative container, we need to either use a
+// container that has stable addressing (true of all std associative
+// containers), or make the key a `std::unique_ptr<CBlockIndex>`
```
Nice doc.

jamesob reviewed (approved):
ACK c68516d662687699b399f8d12e22f2e985edae1a ([`jamesob/ackr/24050.2.dongcarl.validation_move_cblockin`](https://github.com/jamesob/bitcoin/tree/ackr/24050.2.dongcarl.validation_move_cblockin))

Still looks good. Re-reviewed the code, built new revision, ran unittests.


<details><summary>Show signature data</summary>
<p>

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

ACK c68516d662687699b399f8d12e22f2e985edae1a ([`jamesob/ackr/24050.2.dongcarl.validation_move_cblockin`](https://github.com/jamesob/bitcoin/tree/ackr/24050.2.dongcarl.validation_move_cblockin))

Still looks good. Re-reviewed the code, built new revision, ran unittests.

-----BEGIN PGP SIGNATURE-----

iQIzBAEBCgAdFiEEGNRVI1NPYuZCSIrGepNdrbLETwUFAmIEB80ACgkQepNdrbLE
TwX3sQ/+LiR9iKtlAAsH0iGxa5z5pm8BLqTba/1x6IznqZJaC1iztDFRSO+6G6bP
/q3paKXcBLgTBH8n4UQ/ymhmVPLrn4wTsPopDI9ygTOkXMPk3GUfVTqm64dPNaf6
CKMYffWzLsXI3qWUAm6HU3weEwaXNoX1O61rWzsIgrJhrJOX+GhQ6ph1JxMImBjO
4QHcRIh+cMoLEv1fajatdLfDKR8JSW9HDfWcB2n3kLqiZ/jbdN8PhagmZaDkS4b1
lWkTq+m5fpZA1qHQfUknWH+zamuPNorpBnahWmgs8G4bAxEKFeNJMq8a5KZBdy0G
hn1VNMyshtXnxNS4JkoxS+7hv4NBeIaSspwjChA26B18CvIoeRw/2vEKdyEEIJVH
A59cPIHL4hYe7ZJRxC756GzaReYGHS0Myz5FBJ4xEp6/lJbBIr8lrWY3oW7porrv
l50sH4L8ryk2Rhn5qHFhKZ7V1vPSJPLppnzFYyJSzQQlSSqq8NFddJHUv1oz5N2M
wHsRE0FpGlUzTlsn/iAyQdfML5TG1ZIHZvyBHSvbDFeLrvfawzeMRoUf/aSsE+T7
q1YNQf2cNHdxYUIFf+NoeYDZRfH81a71krQy2SuKI/WFDgn57EexQB2Qi8lhBSiG
r3jFuvFvWSihTJQ2Hqt5xbRfLRTgNbjJgNpPjRd+g/yI2rztmEI=
=GA3x
-----END PGP SIGNATURE-----

```

</p></details>

<details><summary>Show platform data</summary>
<p>

```
Tested on Linux-5.10.0-11-amd64-x86_64-with-glibc2.31

Configured with ./configure LDFLAGS=-L/home/james/src/bitcoin/db4/lib/ CPPFLAGS=-I/home/james/src/bitcoin/db4/include/ CXXFLAGS=-fPIE -pipe -O2 -g -Wthread-safety-analysis -Wall -Werror=sign-compare -Wsign-compare -Werror=thread-safety-analysis --enable-wallet --enable-debug --with-daemon --enable-natpmp-default CXX=/usr/local/bin/clang++ CC=/usr/local/bin/clang PYTHONPATH= --disable-shared --with-pic --enable-benchmark=no --enable-module-recovery --enable-module-schnorrsig --enable-experimental --disable-openssl-tests --disable-shared --with-pic --enable-benchmark=no --enable-module-recovery --enable-module-schnorrsig --enable-experimental --disable-openssl-tests --no-create --no-recursion

Compiled with /usr/bin/ccache /usr/local/bin/clang++ -std=c++17 -mavx -mavx2 -mpclmul -fPIE -pipe -O2 -g -Wthread-safety-analysis -Wall -Werror=sign-compare -Wsign-compare -Werror=thread-safety-analysis -O0 -g3 -ftrapv -fdebug-prefix-map=$(abs_top_srcdir)=.  -Wstack-protector -fstack-protector-all -fcf-protection=full -fstack-clash-protection -msse4 -msha -msse4.1 -msse4.2  i

Compiler version: Debian clang version 13.0.1-++20211215043002+d904698b53e4-1~exp1~20211215043040.32
```

</p></details>

MarcoFalke commented:
Assigned 24.0, as this missed the 0.23 feature freeze

dongcarl commented:
Pushed c68516d662687699b399f8d12e22f2e985edae1a -> 6c23c415613d8b847e6f6a2f872be893da9f4384
- Added `const` version of `BlockManager::LookupBlockIndex` to account for changes in fbab43f169924c681ef085639b3e4de6c74a4958
- Removed a now-unnecessary `CBlockIndex` forward declaration

maflcko reviewed (approved):
re-ACK 6c23c415613d8b847e6f6a2f872be893da9f4384 🎨

<details><summary>Show signature</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

re-ACK 6c23c415613d8b847e6f6a2f872be893da9f4384 🎨
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUjQHQwArZhANo4eHc9fHY/ogxBargWcUs82Jt+5QFFFQFB33PIjOCYNE1fpaPuf
y7SVbEdp0SvsMs2qOn86LptVA2UJQYaaYB+36/AlPo0f1tXRHmNZbFSnQicRGnYG
8Mf4S/NUS81e9Mg85rqXjTaCxeAkVtALazna7+wSE1xHcPx9VMy20OcBRye+BzkK
UmzVHIrtjA6/1lVWa7QBWo8S2se6l3GPIT7TARSuH1+tGGH3LRxJ47S5qYbTJciB
Ec1YBWiepcnq1QywMC9GXYIrYZmKPikmRhrpAIKyL3yQh76eksIFFg1qFLq6yD1d
QhKWt5PA4peyGYLhQYh6UUV1YDh5cXM1PEOur17tG6BPw3oX5+j8OsdVIfYur3V5
NM29D2T72nxaBYRrnXH+tIeY6Qz4PN+jKaZSqI3Qa8iaPounJnoryr8FN5SfKIS6
NYCmRCN/EYFC5TocfoWaEue6F69q4Y9AzWE20ko0UF2zOBtFgkv1L/+NDusi+TTc
B8Ozf9u4
=iXAq
-----END PGP SIGNATURE-----
```


</details>
